Version source tagging (automatic and manual) for version_added and deprecation versions

felixfontein opened the pull request:
##### SUMMARY
Extension of #69631, full implementation of ansible/proposals#178.

This PR extends the automatic version_added tagging from #69631 to automatic version number tagging for all version numbers appearing in documentations (plugins and modules). (I used @abadger's [schemas](https://github.com/ansible-community/ansibulled/tree/master/ansibulled/schemas), so I think the PR covers every version appearing.) It also modifies `Display.deprecated()` to show this information in a nicer way:
```
This feature will be removed in version 2.0.0 of community.general.
This feature will be removed in version 2.14 of Ansible-base.
```
It also adds tags to all `module.deprecate()` and `Display.deprecated()` calls, and all `removed_in_version` usages (none) and `deprecated_aliases` usages (one: lib/ansible/module_utils/urls.py) in ansible-base. Finally, it updates validate-modules and pylint's deprecated tests in ansible-test to make sure that versions are correctly tagged, and that all version comparisons in these tests check the tag. (The pylint tests also ensure that the correct tag is used. That's unfortunately not possible in validate-modules, since there only the resulting argument spec can be analyzed.)

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
core

felixfontein committed:
Track collection for version_added.
Validate *all* version numbers in validate-modules.
For tagged version numbers (i.e. version_added), consider source collection to chose validation.

felixfontein committed:
Make tagging/untagging functions more flexible.

felixfontein committed:
Tag all versions in doc fragments.

felixfontein committed:
Tag all deprecation versions issued by code.

felixfontein committed:
Make Display.deprecated() understand tagged versions.

felixfontein committed:
Extend validation to enforce tagged version numbers.

felixfontein committed:
Tag versions in tests.

felixfontein commented:
If I should update ignore.txt instead of adjusting the modules in test/support/integration/plugins/modules/, please tell me.

felixfontein committed:
Lint and fix test.

felixfontein committed:
Mention collection name in collection loader's deprecation/removal messages.

felixfontein committed:
Fix error IDs.

felixfontein committed:
Handle tagged dates in Display.deprecated().

felixfontein commented:
I guess the main thing missing now is tagging deprecation dates. I've updated the display code to handle tagged deprecation dates (`community.general:2020-07-01`), but the sanity tests will scream at them.

felixfontein committed:
Also require that removed_at_date and deprecated_aliases.date are tagged.

felixfontein commented:
The latest version also forces tagged removal dates in the sanity tests. There's no auto-tagging of removal dates yet because there no removal dates in doc fragments and YAML documentation yet - the first step in that direction happens in #69727, once that is merged I'll update the PR (or vice versa).

Please note that deprecation of plugin (i.e. non-module) options by date is not possible right now. (https://github.com/ansible/ansible/blob/devel/lib/ansible/cli/__init__.py#L97-L106)

nitzmahone reviewed (approved):
I'm kinda iffy about tagging all the non-plugin deprecation versions with `ansible.builtin`, but the conversion to ansible-base in the message rendering makes it workable...

felixfontein commented:
@nitzmahone alternatively `ansible-base` would also work, but I thought it might be nicer to stick to something more collection-like so that it's easier to have a more strict validation (which makes sure that the tag is of the form `<python_identifier>.<python_identifier>`), without having to care about the `ansible-base` special case ;-)

(I don't mind changing it if people want that though.)

nitzmahone commented:
I think it's ok for now- it's all internal, so easy to change later

felixfontein commented:
True, except that `ansible-doc` users will see it (for example `ansible-doc community.crypto.openssl_privatekey` shows `version_added: ansible.builtin:2.3` for `attributes`).

nitzmahone commented:
That's fine for those- I was thinking more about the non-plugin-related deprecation cases (the imperative runtime calls to display.deprecated in various Ansible code)

felixfontein committed:
Also automatically tag/untag removed_at_date; fix sanity module removal version check.

felixfontein committed:
Improve error message when invalid version number is used (like '2.14' in collections).

felixfontein commented:
I rebased this PR against the branch of #69727. (There were several conflicts, as expected.)

I also added two new commits:
 - abed542a3aabe1f6873ab0bd9cd0411bb7f48de7 makes sure that the deprecation date added in #69727 is automatically tagged (and correctly validated);
- fd801503faede3a29eb3dd59345dd05a13702d9a makes sure that when module deprecation version/date from meta/runtime.yml is compared to the ones from documentation, that tags are stripped from the documentation values before comparison.

mattclay commented on test/lib/ansible_test/_data/sanity/pylint/plugins/deprecated.py:
```diff
@@ -125,17 +154,71 @@ class AnsibleDeprecatedChecker(BaseChecker):
+        collection = matcher.group(1)
+        date_str = matcher.group(2)
+
+        if collection != (self.collection_name or 'ansible.builtin'):
```
Shouldn't use of `ansible.builtin` be limited to cases where `self.collection_name` is `None`?

felixfontein replied:
`collection_name` is either `None` or a "real" collection name, i.e. a truthy value. Or do I understand your comment wrong?

(In case someone passes `--collection-name ''` when manually running `pylint`, it will be treated the same as `None`. But I'm not sure why we would want to support this.)

mattclay replied:
Ah, I misread this.

mattclay commented on test/lib/ansible_test/_data/sanity/pylint/plugins/deprecated.py:
```diff
@@ -125,17 +154,71 @@ class AnsibleDeprecatedChecker(BaseChecker):
+        collection = matcher.group(1)
+        version_no = matcher.group(2)
+
+        if collection != (self.collection_name or 'ansible.builtin'):
```
Shouldn't use of `ansible.builtin` be limited to cases where `self.collection_name` is `None`?

mattclay commented on test/lib/ansible_test/_data/sanity/pylint/plugins/deprecated.py:
```diff
@@ -125,17 +154,71 @@ class AnsibleDeprecatedChecker(BaseChecker):
+        if collection != (self.collection_name or 'ansible.builtin'):
+            self.add_message('wrong-collection-deprecated-version-tag', node=node, args=(version,))
+
+        if collection == 'ansible.builtin':
```
Shouldn't this be checking `self.collection_name` instead?

You could use the following (but the collection block would need to come first then):

```suggestion
        if self.collection_name:
```

felixfontein replied:
No, since `ansible.builtin:2.14` should not trigger an `ansible-invalid-deprecated-version` or `ansible-deprecated-version` error, but only an `wrong-collection-deprecated-version-tag` error.

(If someone uses `ansible.builtin:1.0.0`, they'll get more errors.)

felixfontein commented:
@mattclay @nitzmahone thanks a lot for reviewing and merging!!!
